internal/auth/internal/mappers/trafficpermissionsmapper/traffic_permissions_mapper.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package trafficpermissionsmapper
+
+import (
+	"context"
+	"sync"
+
+	"github.com/hashicorp/consul/internal/auth/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/internal/resource/mappers/bimapper"
+	pbauth "github.com/hashicorp/consul/proto-public/pbauth/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+type TrafficPermissionsMapper struct {
+	lock                     sync.Mutex
+	mapper                   *bimapper.Mapper
+	workloadIdentityToCTPMap map[string]*pbresource.ID
```

Comment:
Do we need this? Since computed traffic permissions and workload identities are name aligned, it feels like we should be able to compute them on demand.

Reply:
this distinguishes real workload identities from placeholders, which will save a lot of resource lookups and spurious reconcile requests (and become more important when namespace/partitions TPs are added, and if prefix TPs are added)

Reply:
Can we just make this a set of workload identities (`map[string]struct{}`) and rename this? The name of this field confused me.

Reply:
This is to keep track of which workload identities we have actually seen. In the bimapper, we cannot distinguish which entries are real and which are placeholders that we have generated, so in a scenario where we have written 100 Traffic Permissions, but not yet written any workload identities, we will issue reconcile requests and do workload identity lookups for all of them. And then if we have a NamespaceTrafficPermission written, we will do it all again.

Reply:
Hmm, thanks. I worry though that at the cost of optimizing this we'll be creating race conditions between mapper and reconcile that will result in missed events. Here is a situation I am thinking of:

1. I create TP1 for identity `foo` - no reconcile gets triggered because no WI yet
2. I create WI `foo` 
3. Reconcile starts
4. I create TP2 for identity `foo`
5. Mapper calls `getCTPForWorkloadIdentity` for `foo` and gets nil because it's not tracked yet
6. Reconcile starts tracking workload identity
7. Reconcile reads traffic permissions from mapper cache (at this point we only have TP1 because mapper hasn't added it to cache yet)
8. Mapper adds `TP2` to cache but returns no requests because of point 5, and so no reconcile gets triggered to add TP2 to computed traffic permissions

Reply:
I think this is fixed by adding lock calls to the public methods in the mapper, and making the private helper methods require a lock. I can split that out into a separate PR for us to debate the tradeoff of that optimization, and just make sure the many extraneous reconcile requests are fine here.

Reply:
It's ok to address and discuss it in a follow-up PR. 

Adding a lock feels like it could make Reconcile hang and wait for a lock when it needs to call mapper, blocking other reconciles. This doesn't feel like a great idea to me, and I'd want to see if this is better performance-wise than just triggering reconcile for each TP even when there is no WI for it instead. In the future, we want controller-runtime to support concurrent reconciles to speed up performance, and so I feel like having more reconciles should hopefully speed up the reconciliation overall. If we add a lock though, we potentially won't be able to see benefits from concurrent reconciles here.